include/zephyr/init.h

```diff
@@ -151,6 +151,39 @@ struct init_entry {
 	__attribute__((__section__(                                           \
 		".z_init_" #level STRINGIFY(prio)"_" STRINGIFY(sub_prio)"_")))
 
+
+/* Designated initializers where added to C in C99. There were added to
+ * C++ 20 years later in a much more restricted form. C99 allows many
+ * variations: out of order, mix of designated and not, overlap,
+ * override,... but C++ allows none of these. See differences detailed
+ * in the P0329R0.pdf C++ proposal.
+ * Note __STDC_VERSION__ is undefined when compiling C++.
+ */
+#if defined(__STDC_VERSION__) && (__STDC_VERSION__) < 201100
+
+/* Anonymous unions require C11. Some pre-C11 gcc versions have early
+ * support for anonymous unions but they require these braces when
+ * combined with C99 designated initializers, see longer discussion in
+ * #69411.
+ * These braces are compatible with any C version but not with C++20.
+ */
+#  define Z_INIT_SYS_INIT_DEV_NULL  { .dev = NULL }
+
+#else
+
+/* When using -std=c++20 or higher, g++ (v12.2.0) reject braces for
+ * initializing anonymous unions because it is technically a mix of
+ * designated and not designated initializers which is not allowed in
+ * C++. Interestingly, the _same_ g++ version does accept the braces above
+ * when using -std=c++17 or lower!
+ * This Zephyr test compares different C++ -std=... values:
+ *
+ *    ./scripts/twister -v -b -p qemu_cortex_a53 -T tests/lib/cpp/
```

Comment:
please drop CLI snippets from here, maybe just reference the test.

Reply:
> please drop CLI snippets from here, maybe just reference the test.

Done!

Small, pure comment change; @cfriedt can you please re-approve?